contributors/design-proposals/storage/csi-migration.md

```diff
@@ -0,0 +1,381 @@
+}
+```
+
+#### Library Versioning
+Since the library will be imported by various components it is imperative that
+all components import a version of the library that supports in-tree driver x
+before the migration feature flag for x is turned on. If not, the TranslateToCSI
+function will return an error when the translation is attempted.
+
+
+### Non-Dynamic Provisioned Volumes (and volumes provisioned before migration)
+In the OperationGenerator at the start of each volume operation call we will
+check to see whether the plugin has been migrated.
+
+For Controller calls, we will call the CSI calls instead of the in-tree calls.
+When CSI specific information is needed from the PV we can call the translation
+library to do a translation of the PV Source to the necessary CSI Source and get
+the information from there. We also foresee that future controller calls such as
+resize and snapshot will use a similar mechanism. All these external components
+will also need to be updated to accept PV's of any source type when it is given
+and use the translation library to translate the in-tree PV Source into a CSI
+Source when necessary.
+
+For Node calls, the VolumeToMount object will contain the in-tree PV Source,
+this can then be translated by the translation library when needed and
+information can be fed to the CSI components when necessary.
+
+Then the rest of the code in the Operation Generator can execute as normal with
+the CSI Plugin and the annotation in the requisite locations.
+
+Caveat: For ALL detach calls of plugins that MAY have already been migrated we
+have to attempt to DELETE the VolumeAttachment object that would have been
+created if that plugin was migrated. This is because Attach after migration
+creates a VolumeAttachment object, and if for some reason we are doing a detach
+with the in-tree plugin, the VolumeAttachment object becomes orphaned.
+
+
+### In-Line Volumes
+In-line controller calls are a special case because there is no PV. In this case
+we will add the CSI Source JSON to the VolumeToAttach object and in Attach we
+will put the Source in a new field in the VolumeAttachment object
+VolumeAttachment.Spec.Source.InlineVolumeSource. The CSI Attacher will have to
+be modified to also check this location for a source before checking the PV
+itself.
+
+#### New Volume Attachment Source Object
+```
+type VolumeAttachmentSource struct {
+    // Name of the persistent volume to attach.
+    // +optional
+    PersistentVolumeName *string 
+
+    // Copy of VolumeSource from volumes in-lined into pods.
+    InlineVolumeSource *CSIPersistentVolumeSource
+}
+```
+
+We need to be careful with naming VolumeAttachments for in-line volumes. The
+name needs to be unique and A/D controller must be able to find the right
+VolumeAttachment when a pod is deleted (i.e. using only info in Node.Status).
+CSI driver in kubelet must be able to find the VolumeAttachment too to get
+AttachmentMetadata for NodeStage/NodePublish.
```

Comment:
There may need to be some code to be added to deal with the security of namespaced secrets that can be different between inline pod volumes and pv based volumes? (rbd secrets behave this way I think)

Reply:
in-tree PV based volumes will translate to csi PV based volumes. and Inline in-tree to CSI in-tree. I don't think we will need to explicitly deal with the differences between the two.